### PR TITLE
Pinpoint Jboss Plugin

### PR DESCRIPTION
**Pull request by suraj-raturi**
@Xylus - Here raising PR for Jboss plugin, Please have a look and let us know if you see any issue

#972 Tested on Jboss EAP version - 6.4.x, We are in process of checking the compatibility with other versions as well.

### CODE REVIEW

File: profiler/src/main/java/com/navercorp/pinpoint/profiler/instrument/PlainClassLoaderHandler.java [hunk 1]
```diff
@@ -102,7 +102,9 @@ private ClassPool createClassPool(ClassLoader classLoader) throws NotFoundExcept
     private Class<?> injectClass0(ClassPool pool, ClassLoader classLoader, String className, ClassLoadingChecker classLoadingChecker) throws NotFoundException, IOException, CannotCompileException, IllegalArgumentException, IllegalAccessException, InvocationTargetException {
         Class<?> c = null;
         try {
-            c = classLoader.loadClass(className);
+            // Using different approach to load classes via classloader
+            // c = classLoader.loadClass(className);
+            c = Class.forName(className, false, classLoader);
```
Comment (emeroad) [inline profiler/src/main/java/com/navercorp/pinpoint/profiler/instrument/PlainClassLoaderHandler.java]: Hello~ suraj-raturi
Why did this code change?
Tell me the reason to change.

Comment (suraj-raturi) [reply profiler/src/main/java/com/navercorp/pinpoint/profiler/instrument/PlainClassLoaderHandler.java]: Hi @emeroad 

Classloader can load classes via both the ways:
1) classLoader.loadClass 
2) Class.forName 

With classLoader.loadClass I was facing linkage error for example - "Caused by: java.lang.LinkageError: loader (instance of  org/jboss/modules/ModuleClassLoader): attempted  duplicate class definition for name: "com/navercorp/pinpoint/plugin/jdbc/mysql/MySqlConstants", while agent was instrumenting class in jboss environment. 

I looked at the alternative ways to load classes in multi-threaded environment as the issue of linkage error has been reported by many (not specifically with Jboss) with classloader.loadClass.

I tried to load class with Class.forName, it worked for me without any issues for Jboss. To be absolutely sure that it won't break anything, I tested it with Tomcat environment as well and it worked fine.

Let me know your thoughts.

Thanks,
Suraj

**Comment by dawidmalina** [comment]
Would be nice to see this plugin with pinpoint 1.6.0

File: profiler/src/main/java/com/navercorp/pinpoint/profiler/instrument/PlainClassLoaderHandler.java (hunk 1 above)
Comment (emeroad) [reply profiler/src/main/java/com/navercorp/pinpoint/profiler/instrument/PlainClassLoaderHandler.java]: Hello.  @suraj-raturi 
I have a good news.
We solved JDKLogger problem in jboss. #2026

We will solved Concurrent ClassLoading problem.

Comment (suraj-raturi) [reply profiler/src/main/java/com/navercorp/pinpoint/profiler/instrument/PlainClassLoaderHandler.java]: Hi @emeroad ,

Thats good news, our approach while working for Jboss plugin was to go for minimum code change in existing Pinpoint project.
So, we choosed to solve the logger problem via defining logmanager property and logmanager jar in Jboss configuration.

Regarding Classloading issue, will be happy to be involve in the discussion and would like to contribute as well, if a more better approach from above code change in PlainClassLoaderHandler can be done.

Thanks

Comment (emeroad) [reply profiler/src/main/java/com/navercorp/pinpoint/profiler/instrument/PlainClassLoaderHandler.java]: Hi @suraj-raturi 
I fixed the JBoss CL issue from #2117
But This PR need to modify setting.
**plugins/jboss/README.md**
 ~~-Djboss.modules.system.pkgs=com.navercorp.pinpoint~~
 -Djboss.modules.system.pkgs=com.navercorp.pinpoint.bootstrap,
                                                   com.navercorp.pinpoint.common,
                                                   com.navercorp.pinpoint.exception

Comment (suraj-raturi) [reply profiler/src/main/java/com/navercorp/pinpoint/profiler/instrument/PlainClassLoaderHandler.java]: Hi @emeroad ,
I have updated the README file.
Thanks

**Comment by emeroad** [comment]
Hi @suraj-raturi 
This PR branch has conflict.
Please resolve conflict.
 Or Could you please give me **maintainer permission**
 - https://github.com/blog/2247-improving-collaboration-with-forks
2b17b4c2-7444-11e6-8959-2f7ed2012d07

**Comment by suraj-raturi** [comment]
Hi @emeroad ,
I have granted permission to allow edits from maintainers.
Thanks

**Comment by Xylus** [comment]
There is currently a bug in our ASM engine - **You must set `profiler.instrument.engine=JAVASSIST` in *pinpoint.config*** when using jboss plugin.

**Pull request merged**